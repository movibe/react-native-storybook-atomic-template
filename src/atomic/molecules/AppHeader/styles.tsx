import { getStatusBarHeight } from 'react-native-status-bar-height'
import Icon from 'react-native-vector-icons/AntDesign'
import Icon2 from 'react-native-vector-icons/MaterialIcons'
import styled from 'styled-components/native'

import { Column, Row } from '../../../atomic/atoms/Flex'
import { Image } from '../../../atomic/atoms/Image'
import { Text } from '../../../atomic/atoms/Text'

export const Title = styled(Text)`
  font-weight: bold;
  font-size: 18px;
  color: ${({ theme }) => theme.text.default.color};
`

export const SubTitle = styled(Text)`
  font-weight: bold;
  font-size: 24px;
  color: ${({ theme }) => theme.text.default.color};
  padding-left: 5px;
`

export const Logo = styled(Image).attrs({
  resizeMode: 'contain',
})`
  width: 64px;
  height: 18px;
`

export const Container = styled(Column)`
  height: ${getStatusBarHeight()}px;
`
export const ContainerTop = styled(Row)`
  height: 30px;
`

export const ContainerSubtitle = styled(Row)`
  padding-top: 10px;
  height: 50px;
`

export const Left = styled.View`
  flex: 1;
  width: 20px;
  align-items: flex-start;
  flex-direction: row;
  justify-content: flex-start;
`
export const Right = styled.View`
  flex: 1;
  width: 20px;
  align-items: flex-end;
  flex-direction: row;
  justify-content: flex-end;
`

export const Body = styled.View`
  flex: 8;
  justify-content: center;
  align-items: center;
  flex-direction: row;
`
export const ArrowBackIcon = styled(Icon).attrs({ name: 'left' })`
  font-size: 20px;
  color: ${({ theme }) => theme.colors.main};
`
export const FilterIcon = styled(Icon2).attrs({ name: 'filter-list' })`
  font-size: 20px;
  color: ${({ theme }) => theme.colors.main};
`
