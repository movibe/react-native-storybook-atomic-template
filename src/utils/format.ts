export const formatPoints = (points: string | number): string => {
  const minus = Number(points) < 0

  let n = Number(points).toString()
  let r = ''
  let x = 0

  for (let i = n.length; i > 0; i--) {
    r += n.substr(i - 1, 1) + (x == 2 && i != 1 ? '.' : '')
    x = x == 2 ? 0 : x + 1
  }
  return `${minus ? '-' : ''}${r.split('').reverse().join('')}`
}

export interface IFlightsStops {
  name: string
  iata: string
  duration: string
}
