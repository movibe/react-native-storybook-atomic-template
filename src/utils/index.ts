export * from './faker'
export * from './qualityAssurance'
export * from './format'
export * from './withTheme'
