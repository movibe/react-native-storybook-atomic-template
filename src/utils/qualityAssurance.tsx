import { Platform } from 'react-native'

interface ItestID {
  accessible?: boolean
  testID: string
  accessibilityLabel?: string
}
const testID = (testID: string): ItestID =>
  Platform.OS === 'android'
    ? {
        testID,
        accessible: true,
        accessibilityLabel: testID,
      }
    : {
        testID,
        accessible: false,
      }

type componentType =
  | 'ActivityIndicator'
  | 'View'
  | 'Lottie'
  | 'ScrollView'
  | 'Modal'
  | 'TextInput'
  | 'Image'
  | 'Text'
  | 'ImageBackground'
  | 'TouchableOpacity'
  | 'Radio'
  | 'Checkbox'

export const assignTestId = (componentType: componentType, componentId: string): ItestID => {
  const id = `${componentType}_${componentId}`
  return testID(id)
}
export const mockTestID = (componentType: componentType, componentId: string): ItestID => assignTestId(componentType, componentId)
