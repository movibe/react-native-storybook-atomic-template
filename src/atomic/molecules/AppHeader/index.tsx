import * as React from 'react'
import { TouchableOpacity } from 'react-native'

import logoImage from '../../../assets/images/logo.svg'
import { Column } from '../../../atomic/atoms/Flex'
import { assignTestId } from '../../../utils'
import * as S from './styles'

export interface IAppHeaderProps {
  testID?: string
  title?: string
  subtitle?: string
  logo?: boolean
  onBack?: () => void
  onFilter?: () => void
}

export const AppHeader: React.FC<IAppHeaderProps> = ({ testID = 'AppHeader', title, subtitle, logo, onBack, onFilter }): JSX.Element => {
  return (
    <S.Container testID={testID}>
      <S.ContainerTop alignItems='center'>
        <S.Left>
          {onBack && (
            <TouchableOpacity onPress={onBack} {...assignTestId('TouchableOpacity', `${testID}_onBack`)}>
              <S.ArrowBackIcon />
            </TouchableOpacity>
          )}
        </S.Left>
        <S.Body>
          {title && <S.Title testID={`${testID}_title`}>{title}</S.Title>}
          {logo && <S.Logo testID={`${testID}_logo`} source={logoImage} />}
        </S.Body>
        <S.Right></S.Right>
      </S.ContainerTop>
      {subtitle && (
        <S.ContainerSubtitle>
          <Column flex={8} justifyContent='center'>
            {subtitle && <S.SubTitle testID={`${testID}_subtitle`}>{subtitle}</S.SubTitle>}
          </Column>
          <Column flex={1} justifyContent='center' alignItems='flex-end'>
            {onFilter && (
              <TouchableOpacity onPress={onFilter} {...assignTestId('TouchableOpacity', `${testID}_onFilter`)}>
                <S.FilterIcon />
              </TouchableOpacity>
            )}
          </Column>
        </S.ContainerSubtitle>
      )}
    </S.Container>
  )
}
