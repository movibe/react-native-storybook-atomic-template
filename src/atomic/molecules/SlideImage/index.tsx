import React, { useState } from 'react'
import { ImageSourcePropType, View } from 'react-native'
import Carousel, { Pagination } from 'react-native-snap-carousel'
import styled from 'styled-components/native'

import { Column } from '../../../atomic/atoms/Flex'
import { Image } from '../../../atomic/atoms/Image'
import { assignTestId } from '../../../utils'

export interface ISlideImageProps {
  testID?: string
  images: Array<ImageSourcePropType>
  width: number
  height?: number
  onPress?: () => void
  autoPlay?: boolean
  firstItem?: number
  showPagination?: boolean
}

const Container = styled(Column)`
  background-color: ${({ theme }) => theme.colors.background};
  border-bottom-width: 1px;
  border-color: ${({ theme }) => theme.colors.disabled};
  width: 100%;
`
export const SlideImage: React.FC<ISlideImageProps> = ({
  testID = 'SlideImage',
  images,
  width,
  onPress,
  height,
  firstItem,
  autoPlay,
  showPagination,
}): JSX.Element => {
  const [activeSlide, setActiveSlide] = useState(firstItem || 0)

  const renderItem = ({ item, index }) => (
    <Column testID={`${testID}_slide_item`} onPress={onPress} key={index}>
      <Image testID={`${testID}_slide_item_image`} source={{ ...item, height: height, width: width }} />
    </Column>
  )

  const onChangeSlide = (index) => {
    setActiveSlide(index)
  }

  return (
    <Container alignItems='center' testID={testID}>
      <Carousel
        {...assignTestId('View', `${testID}_slide`)}
        data={images}
        renderItem={renderItem}
        sliderWidth={width}
        itemWidth={width}
        firstItem={firstItem}
        onSnapToItem={onChangeSlide}
        autoplay={autoPlay}
        activeSlideAlignment='center'
      />
      {showPagination && (
        <View {...assignTestId('View', `SlideImage_pagination`)}>
          <Pagination
            dotsLength={images.length}
            activeDotIndex={activeSlide}
            dotStyle={{
              width: 10,
              height: 10,
              borderRadius: 5,
              backgroundColor: '#828282',
            }}
            inactiveDotStyle={{
              backgroundColor: '#D8D8D8',
            }}
            inactiveDotOpacity={0.4}
            inactiveDotScale={0.6}
          />
        </View>
      )}
    </Container>
  )
}
