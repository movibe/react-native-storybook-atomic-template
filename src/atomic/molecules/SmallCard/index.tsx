import * as React from 'react'
import { ImageSourcePropType, TouchableOpacityProps } from 'react-native'

import { Text as NText } from '../../../atomic/atoms/Text'
import styled from 'styled-components/native'
import { assignTestId } from '../../../utils'

const Container = styled.TouchableOpacity`
  box-shadow: ${(props) => props.theme.boxShadow};
  border-radius: ${(props) => props.theme.borderRadius};
  overflow: hidden;
`

const Background = styled.ImageBackground`
  flex: 1;
  overflow: hidden;
  justify-content: flex-end;
`

const Overlay = styled.View`
  flex: 0.5;
  justify-content: flex-end;
  background-color: rgba(0, 0, 0, 0.5);
`

const Text = styled(NText)`
  color: #fff;
  font-size: 14px;
  font-weight: bold;
  margin: 10px;
`

export interface ISmallCardProps extends TouchableOpacityProps {
  testID?: string
  text?: string
  image: ImageSourcePropType
  gradient?: boolean
  onPress?: () => void
  width?: number
  height?: number
}

export const SmallCard: React.FC<ISmallCardProps> = ({
  testID = 'SmallCard',
  text,
  image,
  gradient,
  width = 120,
  height = 126,
  ...props
}): JSX.Element => {
  return (
    <Container {...assignTestId('TouchableOpacity', testID)} {...props} style={{ width, height }}>
      <Background source={image}>
        {gradient ? (
          <Overlay>
            <Text testID={testID}>{text}</Text>
          </Overlay>
        ) : (
          <Text testID={testID}>{text}</Text>
        )}
      </Background>
    </Container>
  )
}
