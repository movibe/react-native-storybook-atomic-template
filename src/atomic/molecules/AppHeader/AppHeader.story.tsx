import { action } from '@storybook/addon-actions'
import { text } from '@storybook/addon-knobs'
import * as React from 'react'
import { AppHeader } from '../../../atomic/molecules/AppHeader'

export default {
  title: 'Molecules/AppHeader',
}

export const logo = () => <AppHeader logo onBack={action('onBack')} />
export const normal = () => <AppHeader title={text('title', 'Teste')} />
export const onBack = () => <AppHeader onBack={action('onBack')} title={text('title', 'Teste')} />
export const subtitle = () => <AppHeader onBack={action('onBack')} subtitle={text('subtitle', 'Teste')} />
export const subtitleFilter = () => <AppHeader onBack={action('onBack')} onFilter={action('onFilter')} subtitle={text('subtitle', 'Teste')} />
