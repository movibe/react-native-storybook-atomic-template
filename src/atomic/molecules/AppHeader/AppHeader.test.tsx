import * as React from 'react'
import { fireEvent, render } from 'react-native-testing-library'

import { AppHeader, IAppHeaderProps } from '../../../atomic/molecules/AppHeader'
import { mockTestID, StorybookHoc as wrapper } from '../../../utils'

describe('Molecules/AppHeader', () => {
  let props: IAppHeaderProps = {
    testID: 'AppHeader',
    onBack: jest.fn(),
    onFilter: jest.fn(),
  }

  const { toJSON, findByTestId, update } = render(<AppHeader {...props} />, { wrapper })

  const el_container = mockTestID('View', props.testID!).testID
  const el_logo = mockTestID('Image', `${props.testID!}_logo`).testID
  const el_title = mockTestID('Text', `${props.testID!}_title`).testID
  const el_subtitle = mockTestID('Text', `${props.testID!}_subtitle`).testID
  const el_onBack = mockTestID('TouchableOpacity', `${props.testID!}_onBack`).testID
  const el_onFilter = mockTestID('TouchableOpacity', `${props.testID!}_onFilter`).testID

  it(`render component #${el_container}`, async () => {
    const sut = await findByTestId(el_container)
    expect(sut).toBeTruthy()
  })

  it(`render component #${el_logo}`, async () => {
    props.logo = true
    update(<AppHeader {...props} />)
    const sut = await findByTestId(el_logo)
    expect(sut).toBeTruthy()
  })

  it(`render component #${el_title}`, async () => {
    props.logo = false
    props.title = 'Page fake'
    update(<AppHeader {...props} />)
    const sut = await findByTestId(el_title)
    expect(sut).toBeTruthy()
    expect(sut.children).toEqual([props.title])
  })

  it(`render component #${el_subtitle}`, async () => {
    props.logo = false
    props.title = undefined
    props.subtitle = 'Page Fake'
    update(<AppHeader {...props} />)
    const sut = await findByTestId(el_subtitle)
    expect(sut).toBeTruthy()
    expect(sut.children).toEqual([props.subtitle])
  })

  it(`onPress component #${el_onBack}`, async () => {
    const sut = await findByTestId(el_onBack)
    expect(sut).toBeTruthy()
    fireEvent(sut, 'press')
    expect(props.onBack).toBeCalledTimes(1)
  })

  it(`onPress component #${el_onFilter}`, async () => {
    const sut = await findByTestId(el_onFilter)
    expect(sut).toBeTruthy()
    fireEvent(sut, 'press')
    expect(props.onFilter).toBeCalledTimes(1)
  })

  it('should render snapshot', () => {
    expect(toJSON()).toMatchSnapshot()
  })
})
