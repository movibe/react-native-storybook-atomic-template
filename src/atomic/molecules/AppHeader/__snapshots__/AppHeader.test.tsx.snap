// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/AppHeader should render snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "height": "100vh",
      },
    ]
  }
>
  <View
    accessible={false}
    style={
      Array [
        undefined,
        Object {
          "alignItems": undefined,
          "borderColor": undefined,
          "borderWidth": 0,
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": undefined,
          "padding": 10,
        },
      ]
    }
    testID="View_View"
  >
    <View
      accessible={false}
      style={
        Array [
          Array [
            Object {
              "height": 20,
            },
          ],
          Object {
            "alignItems": undefined,
            "borderColor": undefined,
            "borderWidth": 0,
            "flex": undefined,
            "flexDirection": "column",
            "justifyContent": undefined,
            "padding": 0,
          },
        ]
      }
      testID="View_AppHeader"
    >
      <View
        accessible={false}
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderColor": undefined,
              "borderWidth": 0,
              "flex": undefined,
              "flexDirection": "row",
              "justifyContent": undefined,
              "padding": undefined,
            },
            Array [
              Object {
                "height": 30,
              },
            ],
          ]
        }
        testID="View_View"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-start",
                "flexBasis": 0,
                "flexDirection": "row",
                "flexGrow": 1,
                "flexShrink": 1,
                "justifyContent": "flex-start",
                "width": 20,
              },
            ]
          }
        >
          <View
            accessible={false}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="TouchableOpacity_AppHeader_onBack"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": undefined,
                    "fontSize": 12,
                  },
                  Array [
                    Object {
                      "color": "#F1618C",
                      "fontSize": 20,
                    },
                  ],
                  Object {
                    "fontFamily": "anticon",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexBasis": 0,
                "flexDirection": "row",
                "flexGrow": 8,
                "flexShrink": 1,
                "justifyContent": "center",
              },
            ]
          }
        />
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "flexBasis": 0,
                "flexDirection": "row",
                "flexGrow": 1,
                "flexShrink": 1,
                "justifyContent": "flex-end",
                "width": 20,
              },
            ]
          }
        />
      </View>
      <View
        accessible={false}
        style={
          Array [
            Object {
              "alignItems": undefined,
              "borderColor": undefined,
              "borderWidth": 0,
              "flex": undefined,
              "flexDirection": "row",
              "justifyContent": undefined,
              "padding": undefined,
            },
            Array [
              Object {
                "height": 50,
                "paddingTop": 10,
              },
            ],
          ]
        }
        testID="View_View"
      >
        <View
          accessible={false}
          style={
            Array [
              undefined,
              Object {
                "alignItems": undefined,
                "borderColor": undefined,
                "borderWidth": 0,
                "flex": 8,
                "flexDirection": "column",
                "justifyContent": "center",
                "padding": 0,
              },
            ]
          }
          testID="View_View"
        >
          <Text
            accessible={false}
            style={
              Array [
                Object {
                  "color": "#444",
                  "fontSize": 12,
                  "textTransform": "none",
                },
                Object {
                  "color": "#444",
                  "fontSize": 24,
                  "fontWeight": "bold",
                  "paddingLeft": 5,
                },
              ]
            }
            testID="Text_AppHeader_subtitle"
          >
            Page Fake
          </Text>
        </View>
        <View
          accessible={false}
          style={
            Array [
              undefined,
              Object {
                "alignItems": "flex-end",
                "borderColor": undefined,
                "borderWidth": 0,
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
                "padding": 0,
              },
            ]
          }
          testID="View_View"
        >
          <View
            accessible={false}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="TouchableOpacity_AppHeader_onFilter"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": undefined,
                    "fontSize": 12,
                  },
                  Array [
                    Object {
                      "color": "#F1618C",
                      "fontSize": 20,
                    },
                  ],
                  Object {
                    "fontFamily": "Material Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
